### PR TITLE
Add support for DRM programs with MPEG DASH and Widevine encryption

### PR DESCRIPTION
**Pull request by glennguy**
Has been tested on Ubuntu, Windows 64bit and LibreELEC Rpi3. Not tested on Mac or ODroid C2 but the implementation is there.

### CODE REVIEW

File: resources/lib/config.py [hunk 1]
```diff
@@ -29,6 +29,9 @@
 ISSUE_API_AUTH = 'eGJtY2JvdDo1OTQxNTJjMTBhZGFiNGRlN2M0YWZkZDYwZGQ5NDFkNWY4YmIzOGFj'
 GIST_API_URL = 'https://api.github.com/gists'
 
+BRIGHTCOVE_URL = "https://edge.api.brightcove.com/playback/v1/accounts/{0}/videos/ref:{1}"
```
Comment (andybotting) [inline resources/lib/config.py]: Use single quote for consistency

File: addon.xml [hunk 2]
```diff
@@ -2,6 +2,7 @@
   <requires>
     <import addon="xbmc.python" version="2.1.0"/>
     <import addon="script.module.simplejson"/>
+    <import addon="script.module.requests" version="2.9.1"/>
```
Comment (andybotting) [inline addon.xml]: Is the version actually required?

File: resources/lib/config.py [hunk 3]
```diff
@@ -35,3 +35,34 @@
 api_url = 'https://y7mobile.query.yahoo.com/v1/yql'
 oauth_consumer_key = 'dj0yJmk9QWJodDF5WDVnTGhwJmQ9WVdrOU1HODNiVXB0TnpnbWNHbzlNVGc0TWpnMk5UUTJNZy0tJnM9Y29uc3VtZXJzZWNyZXQmeD04Yw--'
 oauth_consumer_secret = '0e4a80fc03b8ff1ed74a68a8dc583e77ff9e279b'
+
+SSD_WV_REPO = "https://github.com/glennguy/decryptmodules/raw/master/"
+WIDEVINECDM_URL = { 'Linuxx86_64': 'https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb',
+                    'Linuxarmv7': 'http://odroidxu.leeharris.me.uk/xu3/chromium-widevine-1.4.8.823-2-armv7h.pkg.tar.xz',
+                    'Linuxarmv7': 'http://odroidxu.leeharris.me.uk/xu3/chromium-widevine-1.4.8.823-2-armv7h.pkg.tar.xz'}
```
Comment (andybotting) [inline resources/lib/config.py]: Should be armv8?

File: resources/lib/wvhelper.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        return False
+
+    if xbmc.getCondVisibility('system.platform.android'):
+        xbmcgui.Dialog().ok('Android not supported',
+                            'Android not supported for DRM playblack')
+        xbmc.log('Android not supported for DRM playback', xbmc.LOGNOTICE)
+        return False
+
+    if not is_widevinecdm():
+        msg1 =  'Missing widevinecdm module required for DRM content'
```
Comment (andybotting) [inline resources/lib/wvhelper.py]: Extra space

File: resources/lib/wvhelper.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                ' module for your system?')
+        print type(msg1)
+        print type(msg2)
+        if xbmcgui.Dialog().yesno(msg1, msg2, msg3):
+            get_widevinecdm()
+        else:
+            return False
+
+    if not is_ssd_wv():
+        msg1 =  'Missing ssd_wv module required for DRM content'
```
Comment (andybotting) [inline resources/lib/wvhelper.py]: extra space

File: resources/lib/wvhelper.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        for chunk in res.iter_content(chunk_size=chunk_size):
+            f.write(chunk)
+            downloaded += len(chunk)
+            percent = int(downloaded*100/total_length)
+            if dp.iscanceled():
+                dp.close()
+                res.close()
+            dp.update(percent)
+    xbmc.log('Download {0} bytes complete, saved in {1}'.format(
+                int(total_length), download_path),xbmc.LOGNOTICE)
```
Comment (andybotting) [inline resources/lib/wvhelper.py]: should have space after comma

File: resources/lib/wvhelper.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+import xml.etree.ElementTree as ET
+import random
+import zipfile
+from urlparse import urljoin
+
+system_ = platform.system()
+arch = platform.machine()
+if arch[:3] == 'arm':
+    arch = arch[:5]
+git_url = config.SSD_WV_REPO
```
Comment (andybotting) [inline resources/lib/wvhelper.py]: Looks like it's only used one, maybe forget about setting a new var for this

**Comment by glennguy** [comment]
Do you worry that if Google lodged a complaint with Github for hosting their binaries against their licensing that the repo may be taken down?

**Comment by andybotting** [comment]
I've pulled as is - but I do think we should host the libs ourselves. We should host them in a separate repo (just in case), but if they get pulled, then we can just host elsewhere or go back to the more complex code.

**Pull request merged**